### PR TITLE
[Example] update PointNet and PointNet++ examples for Part Segmentation

### PR DESCRIPTION
**Pull request by wcyjames**
## Description

+ Update PointNet++ for Part Segmentation
+ Fix bugs in the original codes
+ Support point cloud visualization for Part Segmentation in Tensorboard

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/pointcloud/pointnet/train_partseg.py [hunk 1]
```diff
@@ -39,7 +41,7 @@ def collate(samples):
         shuffle=True,
         drop_last=True)
 
-def train(net, opt, scheduler, train_loader, dev):
+def train(net, opt, scheduler,  train_loader, dev):
```
Comment (hetong007) [inline examples/pytorch/pointcloud/pointnet/train_partseg.py:44]: unnecessary space

File: examples/pytorch/pointcloud/pointnet/README.md [hunk 2]
```diff
@@ -7,13 +7,21 @@ This is a reproduction of the papers
 
 # Performance
 
+## Classification
```
Comment (hetong007) [inline examples/pytorch/pointcloud/pointnet/README.md:10]: Add numbers from the original paper and the reference implementation to these two tables

Comment (hetong007) [inline examples/pytorch/pointcloud/pointnet/README.md:10]: Also, if per-epoch training time numbers are available for ours and the reference implementation, you can also put into the table.

File: examples/pytorch/pointcloud/pointnet/pointnet2.py [hunk 3]
```diff
@@ -114,6 +114,22 @@ def forward(self, edges):
             res = pos
         return {'agg_feat': res}
 
+class RelativePositionMessageMSG(nn.Module):
```
Comment (hetong007) [inline examples/pytorch/pointcloud/pointnet/pointnet2.py:117]: Why do we need a version for MSG?

File: examples/pytorch/pointcloud/pointnet/train_partseg_with_visualization.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
+import torch
```
Comment (hetong007) [inline examples/pytorch/pointcloud/pointnet/train_partseg_with_visualization.py:1]: Can we merge this with `train_partseg.py`? Ideally in the new file, users can control the visualization with some arguments like `--with-visualize`. This makes it much easier to maintain the computation part, i.e. only update one file if needed.

File: examples/pytorch/pointcloud/pointnet/train_partseg.py (hunk 1 above)
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/train_partseg.py]: Fixed

File: examples/pytorch/pointcloud/pointnet/pointnet2_partseg.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        # Feature Propagation layers
+        l2_feat = self.fp3(l2_pos, l3_pos, l2_feat, l3_feat.unsqueeze(1))
+        l1_feat = self.fp2(l1_pos, l2_pos, l1_feat, l2_feat.permute(0, 2, 1))
+        l0_feat = torch.cat([cat_vec.permute(0, 2, 1), l0_pos, l0_feat], 2)
+        l0_feat = self.fp1(l0_pos, l1_pos, l0_feat, l1_feat.permute(0, 2, 1))
+        # FC layers
+        feat = F.relu(self.bn1(self.conv1(l0_feat)))
+        out = self.drop1(feat)
+        out = self.conv2(out)
+        return out
```
[the remark this answers is not in the document]
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/pointnet2_partseg.py]: Fixed

File: examples/pytorch/pointcloud/pointnet/README.md [hunk 6]
```diff
@@ -27,3 +35,12 @@ For point cloud part-segmentation, run with
 ```python
 python train_partseg.py
 ```
+
+## To Visualize Part Segmentation in Tensorboard
```
[the remark this answers is not in the document]
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/README.md:39]: Added an image

File: examples/pytorch/pointcloud/pointnet/README.md (hunk 2 above)
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/README.md:10]: Added numbers from the original paper and from the other GitHub repo

File: examples/pytorch/pointcloud/pointnet/train_partseg_with_visualization.py (hunk 4 above)
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/train_partseg_with_visualization.py]: Merged

File: examples/pytorch/pointcloud/pointnet/pointnet2.py (hunk 3 above)
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/pointnet2.py:117]: Referenced from https://github.com/yanx27/Pointnet_Pointnet2_pytorch/blob/master/models/pointnet_util.py , not mentioned in the paper

File: examples/pytorch/pointcloud/pointnet/README.md (hunk 2 above)
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/README.md:10]: Not sure whether to mention this if it is machine-dependent

Comment (hetong007) [reply examples/pytorch/pointcloud/pointnet/README.md:10]: You can also add your environment config (CPU model, GPU model, RAM size, etc).

File: examples/pytorch/pointcloud/pointnet/train_partseg.py [hunk 7]
```diff
@@ -25,6 +26,7 @@
 parser.add_argument('--num-epochs', type=int, default=250)
 parser.add_argument('--num-workers', type=int, default=4)
 parser.add_argument('--batch-size', type=int, default=16)
+parser.add_argument('--tensorboard', type=bool, default=False)
```
Comment (hetong007) [inline examples/pytorch/pointcloud/pointnet/train_partseg.py:29]: `parser.add_argument('--tensorboard', action='store_true')`

File: examples/pytorch/pointcloud/pointnet/pointnet2.py (hunk 3 above)
Comment (hetong007) [reply examples/pytorch/pointcloud/pointnet/pointnet2.py:117]: My question is essentially: why do we need a `RelativePositionMessageMSG`, if we already have a `RelativePositionMessage`?

File: examples/pytorch/pointcloud/pointnet/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (hetong007) [reply examples/pytorch/pointcloud/pointnet/README.md:39]: Where's the image?

File: examples/pytorch/pointcloud/pointnet/README.md (hunk 2 above)
Comment (hetong007) [reply examples/pytorch/pointcloud/pointnet/README.md:10]: Where are the numbers?

**Comment by hetong007** [comment]
I guess you might have some local changes that are not pushed to remote. Please double check and address the comments.

File: examples/pytorch/pointcloud/pointnet/README.md [hunk 8]
```diff
@@ -27,3 +36,13 @@ For point cloud part-segmentation, run with
...
+![Screenshot](vis.png)
```
Comment (hetong007) [inline examples/pytorch/pointcloud/pointnet/README.md:41]: where's this file?

File: examples/pytorch/pointcloud/pointnet/pointnet2.py (hunk 3 above)
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/pointnet2.py:117]: Slightly different. The orders of the arguments are somehow switched in their implementations.

Comment (hetong007) [reply examples/pytorch/pointcloud/pointnet/pointnet2.py:117]: What impact does it have, and can we just adapt to the original implmentation?

File: examples/pytorch/pointcloud/pointnet/README.md (hunk 2 above)
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/README.md:10]: Fixed

File: examples/pytorch/pointcloud/pointnet/train_partseg.py (hunk 7 above)
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/train_partseg.py]: Fixed

File: examples/pytorch/pointcloud/pointnet/README.md (hunk 8 above)
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/README.md:41]: Fixed(previously ignored by .gitignore)

File: examples/pytorch/pointcloud/pointnet/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/README.md:41]: Fixed

File: examples/pytorch/pointcloud/pointnet/pointnet2.py (hunk 3 above)
Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/pointnet2.py]: Deleted the MSG version.

File: examples/pytorch/pointcloud/pointnet/README.md [hunk 9]
```diff
@@ -7,13 +7,23 @@ This is a reproduction of the papers
...
+| PointNet        | ModelNet40 | Accuracy | 89.2(Official)   | 89.3        | 181.8             | 95.0          |
+| PointNet++(SSG) | ModelNet40 | Accuracy | 92.4             | 93.3        | 182.6             | 133.7         |
+| PointNet++(MSG) | ModelNet40 | Accuracy | 92.8             | 93.3        | 383.6             | 240.5         |
 
+## Part Segmentation
+
+| Model           | Dataset    | Metric   | Score - PyTorch | Score - DGL | Time(s) - PyTorch | Time(s) - DGL |
+|-----------------|------------|----------|-----------------|-------------|-------------------|---------------|
+| PointNet        | ShapeNet   | mIoU     | 84.3            | 83.6        | 251.6             | 234.0         |
+| PointNet++(SSG) | ShapeNet   | mIoU     | 84.9            | 84.5        | 361.7.            | 240.1         |
```
Comment (hetong007) [inline examples/pytorch/pointcloud/pointnet/README.md:22]: `361.7.` -> `361.7`

Comment (wcyjames) [reply examples/pytorch/pointcloud/pointnet/README.md]: Fixed

**Review by hetong007 — APPROVED** [review]

**Pull request merged**